File: packages/test-plugin/ios/Sources/TestPlugin/TestImpl.swift
import Foundation
import UIKit

/**
 Native iOS implementation for the Test plugin.

 This class contains ONLY platform logic and must not:
 - access CAPPluginCall
 - depend on Capacitor
 - perform mapping logic
 */
@objc public final class TestImpl: NSObject {

    // Properties
    private var config: TestConfig?

    // Initializer
    override init() {
        super.init()
    }

    // MARK: - Configuration

    /**
     Applies the plugin configuration.

     This method should be called exactly once during plugin initialization.
     Its responsibility is to translate configuration values into runtime
     behavior (e.g. enabling verbose logging).
     */
    func applyConfig(_ config: TestConfig) {
        self.config = config
        TestLogger.verbose = config.verboseLogging
        TestLogger.debug("Configuration applied:", config.verboseLogging)
    }

    // MARK: - Echo Method

    /**
     Returns the provided value unchanged.

     This method represents a simple synchronous native operation
     and is intentionally side-effect free.
     */
    @objc public func echo(_ value: String) -> String {
        TestLogger.debug("Echoing value:", value)
        return value
    }

    // MARK: - openAppSettings

    /// Opens the system app settings screen.
    func openAppSettings() throws {
        guard let url = URL(string: UIApplication.openSettingsURLString) else {
            throw TestError.unavailable("Settings URL not available")
        }

        guard UIApplication.shared.canOpenURL(url) else {
            throw TestError.unavailable("Cannot open settings")
        }

        UIApplication.shared.open(url)
    }
}

File: packages/test-plugin/android/src/main/java/io/capkit/test/TestImpl.kt
package io.capkit.test

import android.content.Context
import io.capkit.test.utils.TestLogger

/**
 * Platform-specific native implementation for the Test plugin.
 *
 * This class contains pure Android logic and MUST NOT depend
 * directly on Capacitor bridge APIs.
 *
 * The Capacitor plugin class is responsible for:
 * - reading configuration
 * - handling PluginCall objects
 * - delegating logic to this implementation
 */
class TestImpl(
  private val context: Context,
) {
  /**
   * Cached plugin configuration.
   * Provided once during initialization.
   */
  private lateinit var config: TestConfig

  /**
   * Applies the plugin configuration.
   *
   * This method should be called exactly once during plugin load
   * and is responsible for translating configuration values into
   * runtime behavior (e.g. logging verbosity).
   */
  fun updateConfig(newConfig: TestConfig) {
    this.config = newConfig
    TestLogger.verbose = newConfig.verboseLogging
    TestLogger.debug(
      "Configuration applied. Verbose logging:",
      newConfig.verboseLogging.toString(),
    )
  }

  /**
   * Echoes the provided value.
   *
   * This method represents a simple synchronous native operation
   * and is intentionally side-effect free.
   */
  fun echo(value: String): String {
    TestLogger.debug("Echoing value:", value)
    return value
  }

  /**
   * Opens the system app settings screen.
   *
   * Returns a Result<Unit> to explicitly model success or failure.
   */
  fun openAppSettings(): Result<Unit> {
    return try {
      // Actual intent execution is delegated to the Plugin layer
      Result.success(Unit)
    } catch (e: SecurityException) {
      Result.failure(TestError.PermissionDenied("Permission denied"))
    } catch (e: Exception) {
      Result.failure(TestError.Unavailable("Settings not available"))
    }
  }

  companion object {
    /**
     * Account type identifier (example constant).
     */
    const val ACCOUNT_TYPE = "io.capkit.test"

    /**
     * Human-readable account name (example constant).
     */
    const val ACCOUNT_NAME = "Test"
  }
}
